cachi2/core/package_managers/npm.py

```diff
@@ -173,6 +185,190 @@ def get_sbom_components(self) -> list[dict[str, str]]:
+            raise FetchError(f"{vcs_url} is not valid VCS url! {key} is missing.")
+
+    if url.hostname:
+        vcs_url_info["host"] = str(url.hostname)
+    else:
+        raise FetchError(f"{vcs_url} is not valid VCS url! Host is missing.")
+
+    return vcs_url_info
+
+
+def _clone_repo_pack_archive(
+    vcs: str,
+    download_dir: RootedPath,
+) -> RootedPath:
+    """
+    Clone a repository and pack its content as tar.
+
+    :param url: URL for file download
+    :param download_dir: Output folder where dependencies will be downloaded
+    :raise FetchError: If download failed
+    """
+    info = _extract_git_info_npm(vcs)
+    download_path = RootedPath(
+        Path(
+            download_dir,
+            info["host"],  # host
+            info["namespace"],
+            info["repo"],
+            f'{info["repo"]}-external-gitcommit-{info["ref"]}.tgz',
+        )
+    )
+
+    # Create missing directories
+    directory = os.path.dirname(download_path)
+    os.makedirs(directory, exist_ok=True)
+    clone_as_tarball(info["url"], info["ref"], download_path.path)
+
+    return download_path
+
+
+def _get_npm_dependencies(
+    download_dir: RootedPath, deps_to_download: Dict[str, Dict[str, str]]
+) -> Dict[str, RootedPath]:
+    """
+    Download npm dependencies.
+
+    Receives the destination directory (download_dir)
+    and the dependencies to be downloaded (deps_to_download).
+
+    :param download_dir: Destination directory path where deps will be downloaded
+    :param deps_to_download: Dict of dependencies to be downloaded.
+    :return: Dictionary of dependencies with downloaded paths
+    """
+    files_to_download: dict[str, dict[str, Any]] = {}
+    download_paths = {}
+    for dependency, info in deps_to_download.items():
+        if (
+            info["url"].startswith("github:")
+            or info["url"].startswith("gitlab:")
+            or info["url"].startswith("bitbucket:")
+        ):
+            info["url"] = _update_vcs_url_with_full_hostname(info["url"])
+
+        if info["url"].startswith("git+"):
+            # Git source
+            download_paths[dependency] = _clone_repo_pack_archive(info["url"], download_dir)
+        else:
+            if "registry.npmjs.org" == urlparse(info["url"]).hostname:
+                # Tar archive from npm registry
+                archive_name = f'{dependency}-{info["version"]}.tgz'.removeprefix("@").replace(
+                    "/", "-"
+                )
+                download_paths[dependency] = download_dir.join_within_root(archive_name)
+            else:
+                # Tar archive
+                algorithm, digest = ChecksumInfo.from_sri(info["integrity"])
+                download_paths[dependency] = RootedPath(
+                    Path(
+                        download_dir.path,
+                        f"external-{dependency}",
+                        f"{dependency}-external-{algorithm}-{digest}.tgz",
+                    )
+                )
+
```

Comment:
What about missing integrity here? IMO we should fail (but more gracefully than `from_sri(None)`)

For https dependencies, missing checksum is a much bigger issue than for registry dependencies since there's nothing "locking" them anymore - the content downloaded from an arbitray url may change at any time

Reply:
I agree we should fail in this case I tried to write some polite message for user. Maybe I should add there that they might try to run `npm install` with newer npm version.